buildSrc/src/main/java/org/elasticsearch/gradle/PublishPlugin.java

```diff
@@ -50,27 +52,50 @@ private static String getArchivesBaseName(Project project) {
         return project.getConvention().getPlugin(BasePluginConvention.class).getArchivesBaseName();
     }
 
-    /**Configuration generation of maven poms. */
+    /**
+     * Configuration generation of maven poms.
+     */
     private static void configurePomGeneration(Project project) {
-
         TaskProvider<Task> generatePomTask = project.getTasks().register("generatePom");
+        project.getTasks().named(LifecycleBasePlugin.ASSEMBLE_TASK_NAME).configure(assemble -> assemble.dependsOn(generatePomTask));
+        project.getTasks()
+                .withType(GenerateMavenPom.class)
+                .configureEach(
+                        pomTask -> pomTask.setDestination(
+                                (Callable<String>) () -> String.format(
+                                        "%s/distributions/%s-%s.pom",
+                                        project.getBuildDir(),
+                                        getArchivesBaseName(project),
+                                        project.getVersion()
+                                )
+                        )
+                );
+        PublishingExtension publishing = project.getExtensions().getByType(PublishingExtension.class);
+        final var mavenPublications = publishing.getPublications().withType(MavenPublication.class);
+        addNameAndDescriptiontoPom(project, mavenPublications);
+        configurePomWithShadowPlugin(project, publishing);
 
-        maybeConfigure(project.getTasks(), LifecycleBasePlugin.ASSEMBLE_TASK_NAME, assemble -> assemble.dependsOn(generatePomTask));
+        mavenPublications.all(publication -> {
+            // Add git origin info to generated POM files for internal builds
+            BuildParams.withInternalBuild(() -> publication.getPom().withXml(PublishPlugin::addScmInfo));
+            // have to defer this until archivesBaseName is set
+            project.afterEvaluate(p -> publication.setArtifactId(getArchivesBaseName(project)));
+            generatePomTask.configure(
```

Comment:
Can we simplify this to just have `generatePom` depend on all `GenerateMavenPom` tasks?